### ISSUE TITLE
epic: adopt LLM-as-a-Verifier verification scaling across selection, review, and progress signals

### ISSUE THREAD

**Issue by flora131** [flora131/atomic#2261]
## Summary

Atomic decides which candidate wins, whether a review round passes, and whether a loop should stop by asking a model a compound question and reading back one discrete token — `"first" | "second"`, `"pass" | "fail"`, `done: true | false`. *LLM-as-a-Verifier* (arXiv:2607.05391) shows that this discrete-judge shape is the weakest available reading of the same model call, and that three cheap scaling axes each raise pairwise verification accuracy on Terminal-Bench V2 without any training — score granularity `G`, from 73.1% at `G`=1 to 77.5% at `G`=20 (Fig. 4 left; §4.1); repeated evaluation `K`, from 74.7% at `K`=1 to 77.5% at `K`=16 (Fig. 4 middle; §4.2); and criteria decomposition `C`, from 75.2%–76.4% for any single criterion to 78.3% for the three-criterion ensemble (Fig. 4 right; §4.3) — each measured as its own sweep with the other axes held at their own setting, so the paper reports no single end-to-end 73.1% → 78.3% gain and none is claimed here. This epic tracks seven child issues that move Atomic's selection and review machinery from discrete verdicts to graded, decomposed, aggregated scores, and then reuses the resulting scalar as a progress signal for stall detection. Two of the children — #2254 and #2255 — are prompt-and-aggregation changes that need no engine change and carry the paper's largest single measured gain. The remaining five build on them in dependency order, ending with one exploratory item that is blocked until the rest land. The epic itself ships no code; it exists to hold the shared decisions so seven issues do not answer the same question seven ways.

## Evidence from the paper

The paper's central claim is that verification is a scaling axis distinct from generation. A standard LM judge prompts for a score token and takes the argmax, "collapsing the full scoring distribution into a single value" (§3.1). *LLM-as-a-Verifier* instead takes the expectation over the scoring-token distribution (Eq. 3.1: the mean over `C` criteria and `K` repeats of `Σ_g p(v_g) · φ(v_g)`), converts the resulting continuous reward into a pairwise preference through Bradley–Terry (`P(τi ≻ τj) = σ(R(x,τi) − R(x,τj))`, Eq. 3.2), and then scales along three independent axes.

The measured gains that justify the child issues:

| Axis | Finding | Source |
|---|---|---|
| Criteria decomposition `C` | any single criterion 75.2%–76.4%; the three-criterion ensemble 78.3% | Fig. 4 right; §4.3 |
| Score granularity `G` | 73.1% (`G`=1) → 77.5% (`G`=20); SNR 0.775 → 0.799 | Fig. 4 left; §4.1; Table 1 |
| Repeated evaluation `K` | 74.7% (`K`=1) → 77.5% (`K`=16); variance of the `K`-average estimator shrinks as `O(1/K)`, bias unchanged | Fig. 4 middle; §4.2 |
| Ranking budget | round-robin `O(N²)` → PPT `O(Nk)`, exactly `N + k(N−k) + C(k,2)` pairwise verifications | §3.2 |
| Progress proxy | Value-Order Correlation, Spearman 0.848 ± 0.012 on successful trajectories vs 0.769 ± 0.016 on failed, gap +0.079 | Table 6 |

Each of the first three rows is a separate sweep, run with the other two axes held at their own setting. The paper reports no fixed-budget trade among them and no combined end-to-end figure, so the rows must not be added together, and the span from the granularity floor (73.1%) to the criteria ceiling (78.3%) is a composed reading of two different sweeps rather than a single measurement. Children that reason from a ratio across two axes — #2254 compares points-per-multiple-of-spend for `C` against `K` — must carry that caveat with the comparison.

Two findings frame the whole effort rather than any single child. First, the discrete judge is not a neutral baseline — it is measurably degenerate: it ties 26.7% of comparisons at `K`=1 and still 5.5% at `K`=16, while the verifier yields zero ties at every `K` (Fig. 7 right). On the `query-optimize` case study the discrete 1–5 judge tied 88 of 100 evaluations; taking the expectation over that *same* 1–5 scale eliminated ties and ranked the correct trajectory first in 69 of 100; `G`=20 reached 77 of 100 (Table 2). Second, ensembling a discrete judge does not rescue it: the judge at `K`=16 reaches 74.7%, exactly what the verifier reaches at `K`=1 (Fig. 7 left; §4.2). That is the honest cost bound this epic is working against — sixteen judge calls buy what one verifier call buys.

The selection payoff is bounded by oracle headroom. On Terminal-Bench V2, pass@1 is 83.1% and oracle pass@5 is 92.1%, of which *LLM-as-a-Verifier* captures 86.5% (Table 3). Deriving the capture rate from those three published figures: the headroom is 92.1 − 83.1 = 9.0 points, the realized gain is 86.5 − 83.1 = 3.4 points, so the verifier recovers 3.4 / 9.0 ≈ 38% of the available headroom. On SWE-Bench Verified the pool was heterogeneous — one trajectory each from Claude Opus 4.5, Gemini 3 Flash, and MiniMax M2.5 — with mean pass@1 76.1%, oracle 84.4%, and verifier 78.2% (§5.2, Table 3); the verifier beat every individual contributor to the pool.

## Current behavior in this repository

Atomic's three decision points are each the discrete-judge shape the paper measures against.

**Pairwise selection.** `packages/workflows/builtin/tournament-runner.ts:17-21` defines `judgeDecisionSchema` with `winner: Type.Union([Type.Literal("first"), Type.Literal("second")])` alongside `rationale` and `evidence`. There is no tie and no margin: a coin-flip and a confident call are the same value. `packages/workflows/builtin/tournament-prompts.ts` compounds the problem at the prompt layer — `renderPairwiseJudgePrompt` asks one question spanning "correctness, completeness, evidence, and task fit", which is precisely the compound question §4.3 says verifiers answer by latching onto whichever factor is most salient.

**Bracket structure.** `packages/workflows/builtin/tournament-runner.ts:101-159` runs single-elimination — `while (entrants.length > 1)`, odd counts giving the last entrant a bye — so `N` entrants cost `N−1` judgments and one bad judgment permanently eliminates a candidate. Positional bias is handled by a single line, `const reverse = (round + index) % 2 === 0;` at `packages/workflows/builtin/tournament-runner.ts:111` and again at `:139`, which fixes one orientation per match but does not balance A/B exposure per candidate.

**Review gating.** `packages/workflows/builtin/adversarial-verification-runner.ts:8-13` defines `verifierSchema` with `verdict: Type.Union([Type.Literal("pass"), Type.Literal("fail")])`. The rubric is one 5-bullet markdown file handed whole to every verifier (`packages/workflows/builtin/adversarial-verification-runner.ts:50`) — a monolithic criterion. The gate at `packages/workflows/builtin/adversarial-verification-runner.ts:69` is a unanimity AND: `const allVerifiersPassed = validReports.length === ctx.inputs.verifier_count && validReports.every((report) => report.verdict === "pass");`. And because `validReports` at `:68` filters out anything that failed the structured-output type guard, a *parse failure* is silently counted as a substantive objection.

**Other discrete gates.** `packages/workflows/builtin/generate-and-filter-runner.ts:8-12` carries set-valued shortlists with no per-candidate score, and `packages/workflows/builtin/loop-until-done-runner.ts:18` gates on `done: Type.Boolean()` — a stop condition with no progress magnitude.

**Stall detection has no semantic input.** `packages/subagents/src/runs/shared/subagent-control.ts:16-17` defaults `needsAttentionAfterMs: 60_000` and `activeNoticeAfterMs: 240_000`; `packages/subagents/src/runs/shared/long-running-guard.ts:124` reads `if (metrics.now - metrics.startedAt >= config.activeNoticeAfterMs) return "time_threshold";`. Every existing signal is wall-clock, turn-count, or failure-count. Nothing observes whether the run is making semantic progress — the gap Table 6's VOC result speaks to.

**No logprob plumbing exists.** `grep -rl "logprob\|top_logprobs" packages/` returns nothing. Model access runs through the coding-agent layer inherited from upstream pi (`packages/coding-agent/package.json`), and Anthropic exposes no token logprobs at all. This is why the exact form of Eq. 3.1 is confined to one exploratory child and why everything else must work as a sampling approximation.

## Proposed change

Seven child issues. The dependency edges below are the complete set; nothing else is implied.

```mermaid
graph TD
  A["#2254"] --> C["#2256"]
  B["#2255"] --> C
  C --> D["#2257"]
  C --> E["#2258"]
  C --> G["#2260"]
  D --> G
  F["#2259"]
```

#2254, #2255, and #2259 each declare `depends-on: none`. They are mutually independent: any of them may land first, or concurrently. This epic states no merge order, no rebase requirement, and no sequencing constraint beyond the edges drawn above — each child's own acceptance criteria govern.

Where sequencing language is unavoidable, read **Phase 1** as "the issues with no incoming edge" and **Phase 2** as "everything downstream of #2256". These are labels for the graph, not a schedule.

### Tracked issues

- [ ] #2254 — split the monolithic rubric at `packages/workflows/builtin/adversarial-verification-runner.ts:50` and the compound question in `packages/workflows/builtin/tournament-prompts.ts` into independent per-criterion evaluations, averaged. The paper's code triad is Specification / Output / Errors (§4.3). Largest measured single gain: 75.2–76.4% → 78.3%. No engine change, no logprobs.
- [ ] #2255 — replace the unanimity AND and the parse-failure-as-objection defect at `packages/workflows/builtin/adversarial-verification-runner.ts:68-69` with a mean-score-against-threshold rule, retaining unconditional veto for genuine P0 safety criteria. No engine change.
- [ ] #2256 — replace `winner: "first" | "second"` and `verdict: "pass" | "fail"` with an integer score per criterion, deriving preference through Bradley–Terry (Eq. 3.2), and define what happens at a near-zero margin instead of silently picking.
- [ ] #2257 — replace the single-elimination bracket with the Probabilistic Pivot Tournament: ring pass over a seeded random Hamiltonian cycle, top-`k` pivot selection, pivot rounds, final selection by count-normalized win mass `argmax wᵢ/cᵢ`.
- [ ] #2258 — a periodic scalar progress estimate over a stage's trajectory prefix, consumed by subagent attention escalation, `loop-until-done` stop conditions, ralph's repair-vs-reject gate, and a live gauge. The only child that touches the runtime rather than the workflow library.
- [ ] #2259 — let `tournament` and `generate-and-filter` draw candidates from different models, and document the oracle-headroom routing argument.
- [ ] #2260 — exploratory; surface scoring-token logprobs to a score-only stage mode so Eq. 3.1 can be computed exactly. Blocked until its dependencies land.

### Shared decisions this epic owns

These belong to more than one child. A child that answers one of them differently is creating a contradiction with a sibling, not offering an improvement.

- **One score scale.** Graded scores are integers 1–20 inclusive, matching the paper's `<score_A>` / `<score_B>` prompt (§3.2). `SCORE_MIN = 1`, `SCORE_MAX = 20`, and the normalizer to `[0,1]` live in `packages/workflows/builtin/verification-score.ts` (new), which #2256 creates and owns; #2257, #2258, and #2260 import them. No second scale anywhere, and these constants do not go in `packages/workflows/builtin/review-convergence.ts`, which owns convergence predicates (`findingBlocksClosure`, `MAX_BLOCKING_PRIORITY = 2`, `traceabilityProvenExceptFinalAction`) and nothing else.
- **`verifier_count` keeps its present meaning** — the number of independent repeats, the paper's `K`. Decomposition arrives as a separate input naming the criteria, so total verifier invocations become `criteria × verifier_count`. #2254 states that cost change explicitly.
- **One judge invocation per criterion.** Each invocation sees exactly one criterion and returns one graded score for it. A single call returning several scores reintroduces the compound-question bias §4.3 identifies.
- **Aggregation has one home** — `packages/workflows/builtin/verification-aggregate.ts` (new), created and owned by #2255, covering the threshold, the parse-failure policy, the parse quorum, the per-criterion floor, and the veto path. #2256 changes what each verifier *returns* and adds no second aggregation path.
- **`veto_findings` survives.** #2255 adds it as the retained unanimity path for genuine P0 safety criteria; #2256 reshapes that schema and carries the field through unchanged. Deleting it removes the safety argument that justifies relaxing the AND.
- **Siblings are named by full slug in backticks**, never by issue number.

## Acceptance criteria

- [ ] All seven child issues exist and are linked from the task list above, each carrying the `depends-on:` value recorded in its own metadata fence.
- [ ] Every child issue is closed.
- [ ] `packages/workflows/builtin/verification-score.ts` (new) exists and is the only file in `packages/` that defines a verification score range; `grep -rn "SCORE_MAX" packages/` returns matches in that file and in importers only.
- [ ] `packages/workflows/builtin/verification-aggregate.ts` (new) exists and is the only module that turns a set of verifier results into a round outcome; no second aggregation entrypoint appears under `packages/workflows/builtin/`.
- [ ] `grep -rn "verification-score" packages/workflows/builtin/review-convergence.ts` returns nothing — convergence predicates and score constants stay separate.
- [ ] `npm run check` passes (biome + `tsc --noEmit` + shrinkwrap check).
- [ ] `npm run test:unit`, `npm run test:integration`, and `npm run test:ci-contracts` all pass.
- [ ] No `dist/`, `outDir`, or `tsconfig.build.json` was added under `packages/workflows` — it ships raw TypeScript.
- [ ] No code under `packages/` implements an RL reward path (see non-goals).
- [ ] `packages/coding-agent/docs` reflects the shipped behavior changes, and `packages/*/CHANGELOG.md` carries entries under `## [Unreleased]` for user-visible behavior only.

## Test plan

This epic ships no code and therefore has no test file of its own. Its verification is the union of the children's suites plus the cross-cutting invariants the children individually cannot check.

Each child issue names its own test files and assertions; do not attribute a test name to a sibling from this issue. The commands that must be green before the epic closes:

- `npm run test:unit` — the per-child unit suites.
- `npm run test:integration` — end-to-end workflow runs that exercise the changed selection and review paths.
- `npm run test:ci-contracts` — the repository's own structural contracts, including the per-test timeout policy.
- `npm run check` — biome, `tsc --noEmit`, shrinkwrap.

New tests follow the repository convention: vitest as the runner, assertions in `node:assert/strict`, the shared 30000 ms `TEST_TIMEOUT_MS` budget, and an explicit per-test timeout only where the cost is structural and expressed as a named constant. Files under `packages/workflows/builtin/*` use 2-space indentation; files under `packages/subagents/*` use tabs. Source files use `.js` import extensions. Install with `npm ci --ignore-scripts`.

The one invariant no child owns is scale and aggregation uniqueness — that a single score scale and a single aggregation entrypoint exist across the whole set. It is expressed as the grep-shaped acceptance criteria above rather than as a test, because it is a property of the file set rather than of any runtime behavior.

## Risks and non-goals

**Non-goal: reinforcement learning.** Nothing in §7 of the paper is in scope. The dense-reward results for DSRL-SAC and GRPO, and the RoboRewardBench / Robometer / TOPReward / LIBERO / π0 material, assume a training loop and a VLM path. Atomic has neither. These are named here once as an explicit scope exclusion so no child issue reopens them.

**Non-goal: exact computation of Eq. 3.1 in Phase 1.** Without scoring-token logprobs, everything the first six children do is a sampling approximation of the expectation. That is a real limitation and the children must say so where it bites. #2260 is the only item that pursues the exact form, it is exploratory, and it is blocked behind its dependencies.

**Risk: cost.** Decomposition multiplies verifier invocations by the number of criteria, and repeated evaluation multiplies again by `K`. The paper's own numbers set the ceiling on what that buys — a judge at `K`=16 only reaches what a verifier reaches at `K`=1 (Fig. 7 left) — so without logprobs the return on `K` is bounded. #2254 states the concrete cost change; #2257 is the item that buys budget back, cutting `O(N²)` to `O(Nk)`.

**Risk: the axes do not compose on paper.** Each sweep in the table above was measured with the other axes fixed, so no child may promise the sum of two axes, and no child may treat a gain measured on one axis as evidence for a substitution against another. Where a child argues a cost ratio across axes, it must say plainly that the paper measured no such trade.

**Risk: precision/recall trade in review.** Relaxing the unanimity AND at `packages/workflows/builtin/adversarial-verification-runner.ts:69` reduces spurious repair but also makes it possible for a single genuine objection to be averaged away. This is a deliberate trade, not a pure win, and it is why the veto path is retained rather than deleted.

**Risk: the progress signal is weak.** The paper's success-versus-failure VOC gap is +0.079 (Table 6). That is enough for monitoring and prioritization; it is nowhere near enough for an autonomous kill switch. #2258 must escalate to a human and must degrade to today's wall-clock behavior when the scorer fails.

**Risk: upstream merge cost.** #2260 touches a layer inherited from upstream pi, so it carries ongoing merge cost that the workflow-library children do not.

## References

- *LLM-as-a-Verifier: A General-Purpose Verification Framework* — Kwok, Li, Atreya, Liu, Jiang, Finn, Pavone, Stoica, Mirhoseini. arXiv:2607.05391v2 [cs.AI], 7 Jul 2026. https://arxiv.org/pdf/2607.05391
- Workflow authoring API: `packages/coding-agent/docs/workflows.md`
- Children: #2254, #2255, #2256, #2257, #2258, #2259, #2260

**Comment by flora131** [comment]
Epic superseded by the consolidated set decided in review: #2487 (shared criteria format + aggregation, was #2254+#2255), #2488 (soft-scored tournament in place, was #2256+#2257+pool half of #2259), #2489 (progress-scoring primitive, was #2258), #2490 (goal/ralph graded-signal + convergence-trend adoption, new), #2491 (custom-workflow authoring docs, was the docs half of #2259). #2260 closed as not planned — sampling approximation chosen deliberately. #2200 remains open as the observability substrate.
